File: MyAFBase/MyAFBase/Utils/ReadinessStatus.swift
import Foundation
import SwiftUI

enum ReadinessStatus: Equatable {
    case notSet
    case overdue
    case dueSoon
    case onTrack
    case windowOpen

    var label: String {
        switch self {
        case .notSet: "Not set"
        case .overdue: "Overdue"
        case .dueSoon: "Due soon"
        case .onTrack: "On track"
        case .windowOpen: "Window open"
        }
    }

    var color: Color {
        switch self {
        case .notSet: AppTheme.muted
        case .overdue: AppTheme.danger
        case .dueSoon: AppTheme.warning
        case .onTrack: AppTheme.success
        case .windowOpen: AppTheme.info
        }
    }

    static func evaluate(dueDate: Date?, relativeTo now: Date = .now) -> ReadinessStatus {
        guard let dueDate else { return .notSet }

        let calendar = Calendar.current
        let startOfToday = calendar.startOfDay(for: now)
        let startOfDue = calendar.startOfDay(for: dueDate)

        if startOfDue < startOfToday {
            return .overdue
        }

        guard let soonThreshold = calendar.date(byAdding: .day, value: 30, to: startOfToday) else {
            return .onTrack
        }

        if startOfDue <= soonThreshold {
            return .dueSoon
        }

        return .onTrack
    }

    static func evaluatePCSWindow(start: Date?, end: Date?, relativeTo now: Date = .now) -> ReadinessStatus {
        guard let start, let end else {
            if start != nil || end != nil {
                return .dueSoon
            }
            return .notSet
        }

        let calendar = Calendar.current
        let today = calendar.startOfDay(for: now)
        let windowStart = calendar.startOfDay(for: start)
        let windowEnd = calendar.startOfDay(for: end)

        if today >= windowStart && today <= windowEnd {
            return .windowOpen
        }

        if today > windowEnd {
            return .overdue
        }

        return evaluate(dueDate: start, relativeTo: now)
    }
}

File: android/app/src/main/java/com/ryanladuca/myafbase/domain/logic/ReadinessStatus.kt
package com.ryanladuca.myafbase.domain.logic

import java.util.concurrent.TimeUnit

enum class ReadinessStatus(val label: String) {
    NOT_SET("Not set"),
    ON_TRACK("On track"),
    DUE_SOON("Due soon"),
    OVERDUE("Overdue"),
    WINDOW_OPEN("Window open");

    companion object {
        fun evaluate(dueMillis: Long?, referenceMillis: Long = System.currentTimeMillis()): ReadinessStatus {
            if (dueMillis == null) return NOT_SET
            val days = TimeUnit.MILLISECONDS.toDays(startOfDay(dueMillis) - startOfDay(referenceMillis)).toInt()
            return when {
                days < 0 -> OVERDUE
                days <= 30 -> DUE_SOON
                else -> ON_TRACK
            }
        }

        fun evaluatePcsWindow(startMillis: Long?, endMillis: Long?, referenceMillis: Long = System.currentTimeMillis()): ReadinessStatus {
            if (startMillis == null || endMillis == null) return NOT_SET
            val now = startOfDay(referenceMillis)
            val start = startOfDay(startMillis)
            val end = startOfDay(endMillis)
            return if (now in start..end) WINDOW_OPEN else evaluate(endMillis, referenceMillis)
        }

        private fun startOfDay(millis: Long): Long {
            val cal = java.util.Calendar.getInstance().apply { timeInMillis = millis }
            cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
            cal.set(java.util.Calendar.MINUTE, 0)
            cal.set(java.util.Calendar.SECOND, 0)
            cal.set(java.util.Calendar.MILLISECOND, 0)
            return cal.timeInMillis
        }
    }
}
